### ISSUE TITLE
Neumann BC fail in presence of multiple element types

### ISSUE THREAD

**Issue by djlittl** [gahansen/Albany#148]
For reasons unknown to me, Albany currently supports only a single Neumann field manager (see the function deref_nfm in Albany_Application.cpp).  Because Neumann field managers are created for a specific element topology, it is not currently possible to run a simulation that includes Neumann BC and multiple element types.  A failure occurs because the evaluators in the Neumann field manager assume array lengths based on the element topology.  If the element topology does not match what is expected, an indexing error occurs.

I think part of the root issue is that there is also only a single Dirichlet field manger, and perhaps when the Neumann BC logic was put into place there was some pattern matching done with what is present for DBC.  I'm guess Albany gets away with a single DBC manager because DBC are applied to nodes, and hence there is no need for the DBC to be aware of element topology.  I assume NBC need to know the element topology so that pressures, etc., can be integrated on element faces.  So there needs to be NBC logic to handle different element topologies.  One solution would be multiple NBC field managers, which appear to have been in place some time in the past.

**Comment by bartgol** [comment]
I agree, DBC do not need topology info, just node ids. NBC is a different story, since topology is needed for quadrature.

However, it appears to me that the Neumann class handles each side separately, meaning resizing/updating the quadrature (sorry, I dislike the word cubature) stuff on every side. That's where the Kokkos::DynRankView comes handy, I think (although I dislike also the concept of dynamic rank views, but that's a different story).
